pkg/runtime/wfengine/workflow.go

```diff
@@ -184,20 +196,54 @@ func (wf *workflowActor) createWorkflowInstance(ctx context.Context, actorID str
 		}
 	}
 
-	// We block (re)creation of existing workflows unless they are in a completed state
-	// Or if they still have any pending activity result awaited.
-	if !created {
-		runtimeState := getRuntimeState(actorID, state)
-		if !runtimeState.IsCompleted() {
-			return fmt.Errorf("an active workflow with ID '%s' already exists", actorID)
-		}
-		if wf.activityResultAwaited.Load() {
-			return fmt.Errorf("a terminated workflow with ID '%s' is already awaiting an activity result", actorID)
+	// orchestration didn't exist and just create it.
+	if created {
+		return wf.createInstance(ctx, actorID, startEvent, state)
+	}
+
+	// orchestration already exists, apply reuse id policy
+	runtimeState := getRuntimeState(actorID, state)
+	runtimeStatus := runtimeState.RuntimeStatus()
+	targetStatusValues := backend.BuildStatusSet(reuseIDPolicy.GetOperationStatus())
+	// if target status doesn't match, fall back to original logic, create instance only if previous one is completed
+	if _, ok := targetStatusValues[runtimeStatus]; !ok {
+		return wf.createIfCompleted(ctx, runtimeState, actorID, state, startEvent)
+	}
+
+	switch reuseIDPolicy.GetAction() {
+	case api.IGNORE:
+		// Log an warning message and ignore creating new instance
+		wfLogger.Warnf("An instance with ID '%s' already exists; dropping duplicate create request", actorID)
+		return nil
+	case api.TERMINATE:
+		// terminate existing instance
+		if err := wf.cleanupOrchestrationStateInternal(ctx, actorID, state, false); err != nil {
+			return fmt.Errorf("failed to terminate existing instance with ID '%s'", actorID)
 		}
-		wfLogger.Infof("Workflow actor '%s': workflow was previously completed and is being recreated", actorID)
+
+		// created a new instance
 		state.Reset()
+		return wf.createInstance(ctx, actorID, startEvent, state)
+	}
+	// default Action ERROR, fall back to original logic
+	return wf.createIfCompleted(ctx, runtimeState, actorID, state, startEvent)
+}
+
+func (wf *workflowActor) createIfCompleted(ctx context.Context, runtimeState *backend.OrchestrationRuntimeState, actorID string, state *workflowState, startEvent *backend.HistoryEvent) error {
+	// We block (re)creation of existing workflows unless they are in a completed state
+	// Or if they still have any pending activity result awaited.
+	if !runtimeState.IsCompleted() {
+		return fmt.Errorf("an active workflow with ID '%s' already exists", actorID)
 	}
+	if wf.activityResultAwaited.Load() {
+		return fmt.Errorf("a terminated workflow with ID '%s' is already awaiting an activity result", actorID)
+	}
+	wfLogger.Infof("Workflow actor '%s': workflow was previously completed and is being recreated", actorID)
+	state.Reset()
+	return wf.createInstance(ctx, actorID, startEvent, state)
+}
 
+func (wf *workflowActor) createInstance(ctx context.Context, actorID string, startEvent *backend.HistoryEvent, state *workflowState) error {
```

Comment:
I'm a little confused about the function name here. It's called `createInstance` but the only thing it does is create a reminder? Can we give this a more specific name, like `scheduleWorkflowStart`?